cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -0,0 +1,491 @@
+  /// @param[in] x The `m` by `n` dense block of values (row-major) to
+  /// add to the matrix
+  /// @param[in] rows The row indices of `x`
+  /// @param[in] cols The column indices of `x`
+  void add(const xtl::span<const T>& x,
+           const xtl::span<const std::int32_t>& rows,
+           const xtl::span<const std::int32_t>& cols)
+  {
+    assert(x.size() == rows.size() * cols.size());
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    const std::int32_t num_ghosts0 = _index_maps[0]->num_ghosts();
+    for (std::size_t r = 0; r < rows.size(); ++r)
+    {
+      // Columns indices for row
+      std::int32_t row = rows[r];
+
+      // Current data row
+      const T* xr = x.data() + r * cols.size();
+
+      if (row < local_size0 + num_ghosts0)
+      {
+        auto cit0 = std::next(_cols.begin(), _row_ptr[row]);
+        auto cit1 = std::next(_cols.begin(), _row_ptr[row + 1]);
+        for (std::size_t c = 0; c < cols.size(); ++c)
+        {
+          // Find position of column index
+          auto it = std::lower_bound(cit0, cit1, cols[c]);
+          assert(it != cit1);
+          assert(*it == cols[c]);
+          std::size_t d = std::distance(_cols.begin(), it);
+          _data[d] += xr[c];
+        }
+      }
+      else
+        throw std::runtime_error("Local row out of range");
+    }
+  }
+
+  /// Get the number of local rows
+  /// @param[in] ghost_rows Set to true to include ghost rows in the
+  /// number of local rows
+  std::int32_t num_rows(bool ghost_rows = false) const
+  {
+    return ghost_rows ? _row_ptr.size() - 1 : _index_maps[0]->size_local();
+  }
+
+  /// Copy to a dense matrix
+  /// @note This function is typically used for debugging and not used
+  /// in production
+  /// @param[in] ghost_rows Set to true to include ghost rows in the
+  /// returned matrix
+  /// @return Dense copy of the matrix
+  xt::xtensor<T, 2> to_dense(bool ghost_rows = false) const
+  {
+    const std::int32_t nrows = num_rows(ghost_rows);
+    const std::int32_t ncols
+        = _index_maps[1]->size_local() + _index_maps[1]->num_ghosts();
+    xt::xtensor<T, 2> A = xt::zeros<T>({nrows, ncols});
+    for (std::size_t r = 0; r < nrows; ++r)
+      for (int j = _row_ptr[r]; j < _row_ptr[r + 1]; ++j)
+        A(r, _cols[j]) = _data[j];
+    return A;
+  }
+
+  /// Communicate ghost row data to the owning ranks
+  void finalize()
+  {
+    finalize_begin();
+    finalize_end();
+  }
+
+  /// Begin communication of ghost row data to owning ranks
+  /// @note Calls to this function must be followed by
+  /// MatrixCSR::finalize_end(). Between the two calls matrix values
+  /// must not be changed.
+  void finalize_begin()
+  {
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    const std::int32_t num_ghosts0 = _index_maps[0]->num_ghosts();
+
+    // For each ghost row, pack and send values to send to neighborhood
+    std::vector<int> insert_pos(_val_send_disp);
```

Comment:
I think the memory allocations should be moved to the constructor. The advantage of requiring an exact sparsity pattern is that we can avoid dynamic memory allocations after initial construction.